### PR TITLE
Transfer array from Singe

### PR DESCRIPTION
**Pull request by BrendanKKrueger**
## PR Summary

An implementation of `std::array` that works on both the CPU and the GPU.

## PR Checklist

- [ ] Any changes to code are appropriately documented.
- [ ] Code is formatted.
- [ ] Install test passes.
- [ ] Docs build.
- [ ] If preparing for a new release, update the version in cmake.

### CODE REVIEW

**Review by mauneyc-LANL — APPROVED** [review]

File: ports-of-call/array.hpp [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+/*!
+std::array is not fully constexpr until C++20.  Currently we are relying on
+"constexpr"-ness to offload to GPUs.  PortsOfCall::array provides a
+constexpr replacement for std::array with the following notable exceptions:
+    - the reverse iterators are not provided because std::reverse_iterator is
+not constexpr until C++17 (could be provided with modest effort)
+
+    - the `at` function is not provided because it bounds checks and throws an
+exception which we cannot do on a GPU (it could be added and made
+non-constexpr)
```
Comment (Yurlungur) [inline ports-of-call/array.hpp:22]: I would suggest providing `at` with `PORTABLE_ALWAYS_THROW_OR_ABORT` which will throw an exception on host, but not device based on the protability strategy. (Requires including `portability.hpp` and `portable_errors.hpp`)

File: ports-of-call/array.hpp [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+    - no std::get (cannot be provided without undefined behavior)
+*/
+
+namespace array_detail {
+
+template <typename T, std::size_t N>
+struct bits {
```
Comment (Yurlungur) [inline ports-of-call/array.hpp:30]: I must be missing something---I don't understand the point of this container struct.

File: ports-of-call/array.hpp [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+  using type = T[N];
```
Comment (Yurlungur) [inline ports-of-call/array.hpp:31]: Would prefer this be `Array_t` rather than `type`

File: ports-of-call/array.hpp [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+  template <typename IntT>
+  PORTABLE_FUNCTION static constexpr T &ref_access(type &array,
```
Comment (Yurlungur) [inline ports-of-call/array.hpp:34]: @dholladay00 should correct me if I'm wrong. Do we need `PORTABLE_FUNCTION` decorators for `constexpr`? Either way I think `PORTABLE_FORCEINLINE_FUNCTION` is better here than `PORTABLE_FUNCTION`.

File: test/test_array.hpp [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+  // Fill the array
+  for (int i = 0; i < N; ++i) {
+    arr[i] = i + 1;
+  }
+
+  // Can we read from it?
+  for (int i = 0; i < N; ++i) {
+    CHECK(arr[i] == i + 1);
+  }
```
Comment (Yurlungur) [inline test/test_array.hpp:20]: Some of these loops should be inside device kernels (e.g., `portable_for`) to ensure it works on device.

File: test/test_portsofcall.cpp [hunk 6]
```diff
@@ -18,6 +18,8 @@
 #define CATCH_CONFIG_RUNNER
 #include "catch2/catch.hpp"
 
+#include "test_array.hpp"
```
Comment (Yurlungur) [inline test/test_portsofcall.cpp:21]: Not a fan of this... We should probably just copy the setup we have in singularity to do multiple test files in individual `cpp` files.

**Review by Yurlungur — CHANGES_REQUESTED** [review]
Big picture I think this is a great addition and would like to see this included. Given discussion with @mauneyc-LANL and @rbberger I think we cna merge without CI actually running, so long as you ran tests locally and we can wait for the great test rewrite from @rbberger .

A few comments/questions/requested changes below. Feel free to push back on anything.

**Comment by Yurlungur** [comment]
Oh also there should be at least a mention this exists in the docs.

File: ports-of-call/array.hpp (hunk 3 above)
Comment (mauneyc-LANL) [reply ports-of-call/array.hpp:31]: `Array_t` is somewhat quixotic; `type` is standard trait naming.

**Comment by mauneyc-LANL** [comment]
I was digging around a bit and found this

https://github.com/NVIDIA/libcudacxx/issues/51#issuecomment-717984614

TLDR: 1) `--expt-relaxed-constexpr` & `-std=c++17` seems to provide sufficient functionality for `std::array` on device, 2.) `libcu++` rolls a `std::array` in `cuda/std/array` (as of 2023).

Would this make a hand-rolled `std::array` necessary?

**Comment by Yurlungur** [comment]
> I was digging around a bit and found this
> 
> [NVIDIA/libcudacxx#51 (comment)](https://github.com/NVIDIA/libcudacxx/issues/51#issuecomment-717984614)
> 
> TLDR: 1) `--expt-relaxed-constexpr` & `-std=c++17` seems to provide sufficient functionality for `std::array` on device, 2.) `libcu++` rolls a `std::array` in `cuda/std/array` (as of 2023).
> 
> Would this make a hand-rolled `std::array` necessary?

:+1: This is what we do in parthenon. But is that fully portable?

File: ports-of-call/array.hpp (hunk 2 above)
Comment (BrendanKKrueger) [reply ports-of-call/array.hpp:30]: No idea -- this was CJ code.  I can only assume that this allows you to not have to handle special cases for `N = 0`?

File: ports-of-call/array.hpp (hunk 4 above)
Comment (BrendanKKrueger) [reply ports-of-call/array.hpp:34]: Several years ago we had some issues where `constexpr` wasn't completely sufficient.  It may have just been a quirk of an older compiler or something like that, and maybe they're redundant now.

Comment (Yurlungur) [reply ports-of-call/array.hpp:34]: no I think the decorator may still be desirable... but since it's inline no reason not to use `PORTABLE_FORCEINLINE_FUNCTION` rather than `PORTABLE_FUNCTION`

**Comment by BrendanKKrueger** [comment]
> > I was digging around a bit and found this
> > [NVIDIA/libcudacxx#51 (comment)](https://github.com/NVIDIA/libcudacxx/issues/51#issuecomment-717984614)
> > TLDR: 1) `--expt-relaxed-constexpr` & `-std=c++17` seems to provide sufficient functionality for `std::array` on device, 2.) `libcu++` rolls a `std::array` in `cuda/std/array` (as of 2023).
> > Would this make a hand-rolled `std::array` necessary?
> 
> 👍 This is what we do in parthenon. But is that fully portable?

_Most_ of `std::array` works on GPUs with relaxed `constexpr`, but not all.  In particular, `std::array::fill` and `std::array::swap` aren't `constexpr` until C++20 and I believe Singe code uses `std::array:fill`.  Once we move to C++20 (hopefully much faster than the 14-to-17 transition), then this code becomes redundant if we're comfortable relying on relaxed `constexpr` to get the STL onto the GPU.

**Comment by Yurlungur** [comment]
> > > I was digging around a bit and found this
> > > [NVIDIA/libcudacxx#51 (comment)](https://github.com/NVIDIA/libcudacxx/issues/51#issuecomment-717984614)
> > > TLDR: 1) `--expt-relaxed-constexpr` & `-std=c++17` seems to provide sufficient functionality for `std::array` on device, 2.) `libcu++` rolls a `std::array` in `cuda/std/array` (as of 2023).
> > > Would this make a hand-rolled `std::array` necessary?
> > 
> > 
> > 👍 This is what we do in parthenon. But is that fully portable?
> 
> _Most_ of `std::array` works on GPUs with relaxed `constexpr`, but not all. In particular, `std::array::fill` and `std::array::swap` aren't `constexpr` until C++20 and I believe Singe code uses `std::array:fill`. Once we move to C++20 (hopefully much faster than the 14-to-17 transition), then this code becomes redundant if we're comfortable relying on relaxed `constexpr` to get the STL onto the GPU.

that's a satisfying explanation for me. :+1:

**Comment by BrendanKKrueger** [comment]
This is ready for another review

File: test/test_array.cpp [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+TEST_CASE("array nominal element access (GPU)", "[array][GPU]") {
+  // Declare the array
+  constexpr int N = 16;
+  array<double, N> arr;
+
+  int count = 0;
+  auto func = PORTABLE_LAMBDA(const int i, int &count) mutable {
+    // Fill the array
+    arr[i] = i + 1;
```
Comment (Yurlungur) [inline test/test_array.cpp:21]: I think the fill has to be done on host but it can be checked on device, because lambda will do a const capture.

**Review by Yurlungur — APPROVED** [review]
Looks great! Only one concern in the tests otherwise I think this is good to go.

File: ports-of-call/array.hpp [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+#define _PORTS_OF_CALL_ARRAY_HPP_
+
+#include "portability.hpp"
+#include "portable_errors.hpp"
+
+#include <cstddef>
+#include <iterator>
+#include <tuple>
+
+namespace PortsOfCall {
```
Comment (BrendanKKrueger) [inline ports-of-call/array.hpp:11]: @Yurlungur, is this namespacing correct?

Comment (Yurlungur) [reply ports-of-call/array.hpp:11]: Yep :+1:
